runtime/runtime_storage.go

```diff
@@ -335,64 +338,157 @@ func (s *runtimeStorage) writeCached(inter *interpreter.Interpreter) {
+	// to be run again, until the batch is empty.
+	batch := items
+	for len(batch) > 0 {
+		// parallelize the encoding for items within the same batch
+		// the encoding has no side effect
+		encodedResults, err := s.encodeWriteItems(ENCODING_NUM_WORKER, batch)
+		if err != nil {
+			panic(err)
+		}
+		var newBatch []writeItem
+		// process encoded items of the batch, play side effect for
+		// each item
+		for i, result := range encodedResults {
+			item := batch[i]
+			newItems, err := s.processEncodedItem(item, result)
 			if err != nil {
 				panic(err)
 			}
+			newBatch = append(newBatch, newItems...)
+		}
+		batch = newBatch
+	}
+}
 
-			for _, deferredValue := range deferrals.Values {
+type encodedResult struct {
+	newData   []byte
+	deferrals *interpreter.EncodingDeferrals
+	index     int // the index of the encoded item in the batch
+	err       error
+}
 
-				deferredStorageKey := StorageKey{
-					Address: item.storageKey.Address,
-					Key:     deferredValue.Key,
-				}
+func (s *runtimeStorage) encodeWriteItems(nWorker int, batch []writeItem) ([]*encodedResult, error) {
+	// cache all the encoded results, including errors
+	results := make(chan *encodedResult, len(batch))
+	defer close(results)
 
-				if !deferredValue.Value.IsModified() {
-					continue
-				}
+	// if the number of items in a batch is less than the default number of workers, we will only create the
+	// same amount of workers as the items
+	if len(batch) < nWorker {
+		nWorker = len(batch)
+	}
 
-				items = append(items, writeItem{
-					storageKey: deferredStorageKey,
-					value:      deferredValue.Value,
-				})
+	// jobs buffers at most nWorker number of jobs for workers to work on in parallel
+	// each job contains the index of the item in the batch
+	jobs := make(chan int, nWorker)
+	defer close(jobs)
+
+	worker := func(jobs <-chan int, results chan<- *encodedResult) {
+		for i := range jobs {
+			item := batch[i]
+			if item.value == nil {
+				results <- nil
+			} else {
+				// TODO: encode value with a context, so that if there 10 jobs to work on,
+				// and one job failed, we could cancel the context and skip the rest of
+				// unprocssed jobs without processing them.
+				newData, deferrals, err := s.encodeValue(item.value, item.storageKey.Key)
+				results <- &encodedResult{
+					newData:   newData,
+					deferrals: deferrals,
+					index:     i,
+					err:       err,
+				}
 			}
+		}
+	}
 
-			for _, deferralMove := range deferrals.Moves {
+	for i := 0; i < nWorker; i++ {
+		go worker(jobs, results)
+	}
 
-				s.move(
-					deferralMove.DeferredOwner,
-					deferralMove.DeferredStorageKey,
-					deferralMove.NewOwner,
-					deferralMove.NewStorageKey,
-				)
-			}
+	// push jobs to the workers.
+	// block if no more worker is available
+	for i := range batch {
+		jobs <- i
+	}
+
+	// initialize the array, so that we can insert the encoded result at
+	// the right position.
+	encodedResults := make([]*encodedResult, len(batch))
```

Comment:
I think this might be O(N) memory for a batch of size `N`. If there are 1000 items in a batch, we will cache 1000 encoded results before processing them. 

The second approach would be to break the batch into smaller bundles, each bundle has up to 10 encoded results, which is the same as the max number of workers. Then we will only buffer up to 10 encoded results.

The second approach is slower than the current approach if the batch size is more than 10. Since it can't work on the next bundle, if the previous bundle hasn't been finished. But still faster than no parallelization if the batch size is bigger than 1.

Thoughts? Should we go for second approach? @turbolent

Reply:
I switched to the second approach. Turns out the change is quite simple. Will post some benchmark data